### PR TITLE
Document Webpack

### PR DESCRIPTION
**Pull request by bjacobel**
Add documentation for using Webpack and reorganize/reword recommendations for RequireJS usage.

### Reviewers

Possible roles follow. The PR submitter checks the boxes after each reviewer finishes and gives :+1:. 

- [x] Subject matter expert: @arizzitano 
- [x] Subject matter expert: @andy-armstrong 
- [x] Doc team review (sanity check, copy edit, or dev edit?): @edx/doc

FYI: Tag anyone else who might be interested in this PR here.

@danielmcquillen 

### Testing

- [x] Ran ./run_tests.sh without warnings or errors

### HTML Version (optional)

- [x] http://draft-webpack-2017-04-18.readthedocs.io/en/latest/user_interface_development.html

### Post-review

- [ ] Add a comment with the description of this change or link this PR to the next release notes task.
- [ ] Squash commits

### CODE REVIEW

**Comment by bjacobel** [comment]
fyi @edx/fedx-team

File: en_us/developers/source/user_interface_development.rst [hunk 1]
```diff
@@ -179,87 +179,36 @@ reasons, targeted changes to CoffeeScript will still be accepted.
...
 
-.. _use_requirejs_to_manage_file_dependencies:
+==================================
+Use Webpack to Bundle Dependencies
+==================================
 
-=========================================
-Use RequireJS to Manage File Dependencies
-=========================================
+Write your JavaScript as a `ES2015 module`_. There should be one root file per
```
Comment (arizzitano) [inline en_us/developers/source/user_interface_development.rst]: nit: `an ES2015 module`?

File: en_us/developers/source/user_interface_development.rst [hunk 2]
```diff
@@ -308,33 +257,44 @@ Add the new library to the list of vendor libraries that are installed by
...
+Use React or Backbone to Build Your New UI
+==========================================
+
+Code written with these frameworks is more modular, more extensible, and more easily
+unit tested with Jasmine. React is suggested framework for all greenfield development.
+Development in Backbone will be deprecated in the future. See the `React rollout plan`_
+for more details.
+
+See `How to add a new feature to LMS or Studio`_ to see how to
+structure your code.
```
Comment (arizzitano) [inline en_us/developers/source/user_interface_development.rst:281]: This feels kind of sparse for now, but we can always update later to point at a React style guide once we have one.

File: en_us/developers/source/user_interface_development.rst [hunk 3]
```diff
@@ -349,13 +309,93 @@ Here are a few examples.
 
 * `Teams topic card`_
 
+==============================================
+Use RequireJS where Webpack is not implemented
+==============================================
+
+Where not possible to use Webpack, `RequireJS`_ can be used to manage loading
```
Comment (arizzitano) [inline en_us/developers/source/user_interface_development.rst]: Might be worthwhile to expand on the scenarios where it's acceptable to not use Webpack (legacy code, etc.) so folks aren't tempted to stick with Require.

**Review by arizzitano — APPROVED** [review]

File: en_us/developers/source/user_interface_development.rst (hunk 3 above)
Comment (bjacobel) [reply en_us/developers/source/user_interface_development.rst]: Check out https://github.com/edx/edx-documentation/pull/1487/commits/90b151268eedb5951e9e8b4d78d482f627a1cb06 ?

File: en_us/developers/source/user_interface_development.rst [hunk 4]
```diff
@@ -308,33 +265,44 @@ Add the new library to the list of vendor libraries that are installed by
...
-.. _use_underscore_and_requirejs_text:
+==========================================
+Use React or Backbone to Build Your New UI
+==========================================
 
-===========================================================
-Use Underscore and RequireJS Text for Client-Side Templates
-===========================================================
+Code written with these frameworks is more modular, more extensible, and more easily
+unit tested with Jasmine. React is suggested framework for all greenfield development.
```
Comment (andy-armstrong) [inline en_us/developers/source/user_interface_development.rst]: Nit: "is THE suggested framework"

**Review by andy-armstrong — APPROVED** [review]
👍 nice clean up!

**Comment by arizzitano** [comment]
:+1: on the changes, LGTM

**Comment by catong** [comment]
@bjacobel LGTM 👍

**Pull request merged**